### ISSUE TITLE
Non-determinism in Masks

### ISSUE THREAD

**Issue by shahidhaider-altis** [wasserth/TotalSegmentator#445]
Hello

I have been using TS v1.5.6 for my work and I am noticing that two runs of the total task without the fast flag, using the python api, on the same series do not produce exactly the same masks. 

Can you (@wasserth ) comment on any potential sources of non-determinism in the results when using the defaults?

Thanks

**Comment by wasserth** [comment]
A deep convolutional neural network implemented in pytorch + cuda contains several sources of non-determinism. This goes down to pytorch and cuda. I am not familiar with all their details, but it is very difficult (and reduces performance) to get them 100% deterministic. Therefore, we have to live with a little bit of variability between runs. This variability is minimal.

**Comment by shahidhaider-altis** [comment]
Thank you for your response @wasserth , however, according to cuda/cudnn documentation (https://docs.nvidia.com/deeplearning/cudnn/backend/latest/developer/misc.html) there should not be non-determinism when rerunning on the same GPU. 

I have another model that is using nnunetv2 and I have no issues with determinism there unless we are running it on a different GPU architecture.

**Comment by shahidhaider-altis** [comment]
Can you highlight the sources of non-determinism?
Because according to this https://docs.nvidia.com/deeplearning/cudnn/backend/latest/developer/misc.html the non-determinism at the base cudnn level is from algos run on the backward pass.

If the reason for the non-determinism in TS is from issues listed here: https://pytorch.org/docs/stable/notes/randomness.html, I tried all these flags and still got different results through the python API

**Comment by wasserth** [comment]
I also played with these flags and it did not change things for me. I guess they might get overwritten somewhere in the nnU-Net code base, but I am not familiar with these details of nnU-Net.

**Comment by shahidhaider-altis** [comment]
It appears that it is something others are seeing as well
https://github.com/MIC-DKFZ/nnUNet/issues/1016
however they are no plans to change anything.
